Implement HDMI Tester functionality [REVPI-3124]

tboehler1 opened the pull request:
Supersedes #1

tboehler1 committed:
Add tests for video and audio output

Roughly, outputs are sent to the framebuffer and ALSA default device to
test the functionality of video and audio output.
A more precise description of how these tests work is given in the
README.

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

tboehler1 committed:
Add LICENSE (GPL2) and SPDX headers to sources

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

iluminat23 reviewed (changes requested):
Please add proper commit messages.

add functionality ist a **bit** short

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,142 @@
+    let frame_thread = thread::spawn(|| -> anyhow::Result<()> { frame() });
+    let siren_thread = thread::spawn(|| -> anyhow::Result<()> { siren() });
+
+    if let Ok(s) = frame_thread.join() {
```
One char variables are not so nice. Give it at least a shred of context.

tboehler1 replied:
I disagree in this case: The variable is used in a one line long block and the context is clearly given by the call to `join`, which returns whatever value the joined thread returns after terminating. Artificially making the variable name longer will introduce noise.
To me this is similar enough to something like using `i` in a for loop that it's okay.

If it's not clear what the variable represents I'll make the change.

iluminat23 replied:
To use i in a for loop is ok in most cases. But for me only if i is incremented only in single steps AND if it is not usfull to name it otherwise. And it is somthing it has a long "tradition" and is clear for everybody to read.
I do not really want to start a new tradition here. I don't even know for what s might be the short form. For status?

tboehler1 replied:
I named the variable to carry more context this time, but I'm still against doing this in the future for this trivial case as it's not idiomatic for Rust to be this explicit when this much context is given already. A shorter form, which is not possible here because `join()` returns a type with specific constraints, would be `frame_thread.join()??`.

iluminat23 replied:
One letter variables are a real pita. In every language. It is not necessary to blow the variable up to `frame_thread_join` but something what is better memorable then `s` would be nice. Something like `status` or `tstat` or what ever.

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,142 @@
+}
+
+/// Sound a siren over the default ALSA device. The siren consists of 2 tones, each being played
+/// back for `AUDIO_LENGTH` seconds.
```
As the tests run concurrently it would be good if the tests take the same time. It is a bit strange that the visual output takes longer then the audio.

tboehler1 replied:
You're right, I'll change that.

iluminat23 commented on rust-toolchain.toml:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: GPL-2.0-only
```
```suggestion
# SPDX-License-Identifier: GPL-2.0-or-later
```

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,148 @@
-fn main() {
-    println!("Hello, World!");
+/*
+ * SPDX-License-Identifier: GPL-2.0-only
```
```suggestion
 * SPDX-License-Identifier: GPL-2.0-or-later
```

iluminat23 commented on Cargo.toml:
```diff
@@ -1,10 +1,19 @@
+# SPDX-License-Identifier: GPL-2.0-only
+# Copyright 2023 KUNBUS GmbH
```
```suggestion
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: Copyright 2023 KUNBUS GmbH
```
We provide our source under the unmodified GPLv2 so we retain the or later part. You also provide the full unchanged GPLv2 in the LICENSE.txt, so the license is to be tagged with GPL-2.0-or-later.
We should also provide the SPDX-FileCopyrightText tag.

tboehler1 replied:
Seems like the `SPDX-FileCopyrightText` tag was lost somewhere - I'll add that.

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,150 @@
+
+    for _ in 0..AUDIO_LENGTH * 44100 / 1024 {
+        assert_eq!(io.writei(&buf[..])?, 1024);
+    }
```
Three times almost the same code. This would qualify for a function or macro or what ever.
The somewhat unclear `a` could also be replaced with something more with more meaning. I think in the alsa context this qualifies as frame. So would could replace `a` by `frame` or at least `f` or anything in between.
https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#gabc748a500743713eafa960c7d104ca6f

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,156 @@
+}
+
+fn main() -> anyhow::Result<()> {
+    if let Some(a) = std::env::args().nth(1) {
+        if a == "--version" {
```
```suggestion
    if let Some(arg) = std::env::args().nth(1) {
        if arg == "--version" {
```
Why not using `arg` as name. It is not much longer and it caries all the information. I like concise but clear identifiers. And I don't think these are against idiomatic rust.

tboehler1 committed:
Refactor by extracting code into functions

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

tboehler1 committed:
Introduce Color enum

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

tboehler1 committed:
Add help text for cli

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

iluminat23 commented on src/main.rs:
```diff
@@ -141,12 +141,22 @@ fn siren() -> anyhow::Result<()> {
+            "-v" => eprintln!("{name}: {version}"),
+            _ => {
+                eprintln!(
+                    "Usage: {name} [-v] [-h]\n\
```
It would be nice to have a short summery of the functionality in the help.

tboehler1 replied:
To make this easy, and because when adding another command line switch I would introduce this nonetheless: Do you think I should introduce `clap` yet to make this easier?

iluminat23 replied:
If you think it would be better, you can use clap. I leave this up to you. As you said if the program would get extended it makes sense to use clap then.

iluminat23 commented on src/main.rs:
```diff
@@ -1,3 +1,178 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ * SPDX-FileCopyrightText: Copyright 2023 KUNBUS GmbH
```
```suggestion
 * SPDX-License-Identifier: GPL-2.0-or-later
 * SPDX-FileCopyrightText: Copyright 2023 KUNBUS GmbH
```
Just to be consistent with the other files

tboehler1 committed:
Expand README for rust-toolchain.toml and Docker

Give another example for `aarch64-unknown-linux-gnu` for the
`.cargo/config.toml` file and expand on `rust-toolchain.toml` and give
an example there, too.
Also mention Docker and that the accompanied Dockerfile can be used to
cross-compile for a different architecture and, possibly, glibc version.

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

tboehler1 committed:
Introduce clap for commandline argument parsing

Parsing command line arguments by hand is a pain that clap helps
alleviate. This also makes expanding on cli arguments later easier.
Flags for both help and version exist with sane defaults.

Signed-off-by: Thomas Böhler <t.boehler@kunbus.com>

iluminat23 commented on src/main.rs:
```diff
@@ -139,24 +142,9 @@ fn siren() -> anyhow::Result<()> {
-        }
-        std::process::exit(0);
-    }
+    // struct itself is not useful yet as no information is passed
```
This comment is a bit miss leading. The parse() function has a purpose. The comment reads like the whole line has only documentation value. Could you add a bit of explanation that this used to probvide the user with the information provided in the toml file.
